File: matlab/ELANFIS.m

tic;
clc;
clear;
close all;

% load dataset
data = csvread('routput.csv',1,0);
input = data(:, 1:end-1);
target = data(:, end);


% Initialize parameter
[center,U] = fcm(input, 3, [2 100 1e-3]); %center = center cluster, U = membership level
[samples, features] = size(input);
class = 3; % total classes
ep = 0;
epMax = 1;
[Yy, Hi] = max(U); % Yy = max value between both membership function, Hi = the class corresponding to the max value

bestMSE = 1000;
besta = zeros(class, features);
bestb = repmat(2, class, features);
bestc = zeros(class, features);

while ep < epMax
    ep = ep + 1;
    a = zeros(class, features);
%     b = repmat(2, class, total_features);
b = zeros(class, features);
    c = zeros(class, features);
    % Estimating random mf parameters
    for k =1:class
        for i = 1:features % looping for all features
            % premise parameter: a
            Rj=max(input(:, i))-min(input(:, i));
            m=sum(Hi' == k);
            aTemp = (Rj)/(2*m-2);
            aLower = aTemp*0.5;
            aUpper = aTemp*1.5;
            a(k, i) = (aUpper-aLower).*rand()+aLower;

            %premise parameter: c
            dcc = (2.1-1.9).*rand()+1.9;
            cLower = center(k,features)-dcc/2;
            cUpper = center(k,features)+dcc/2;
            c(k,i) = (cUpper-cLower).*rand()+cLower;
        end
    end
    
    H = [];
    Mu = zeros(samples, class, features); % Mu: miu all samples (total samples x total classes x total features)
    %Calculate membership grades
    for i = 1:samples % looping for each samples for forward pass
        %Calculate firing strength
        for k = 1:class
            w1(k) = 1; % w (not w bar)
            for j = 1:features
                mu(k,j) = 1/(1 + ((input(i,j)-c(k,j))/a(k,j))^(2*b(k,j))); % mu: miu of one sample
                w1(k) = w1(k)*mu(k,j); % fill w for k-th class
                Mu(i, k, j) = mu(k, j);
            end;
        end;
        %Calculate Normalised Firing
        w = w1/sum(w1); % w = w bar one row / one sample data
        ZX = [];
        %Generate X of f=XZ
        for k = 1:class
            ZX = [ZX w(k)*input(i,:) w(k)];
        end; 
        H = [H; ZX]; % combine matrix H of each sample
    end

    % consequent parameter (p, q, r)
    beta = pinv(H) * target; % moore pseudo invers

    output = H * beta; % calculate output from weight
    
    error=target - output;
    MSE = mean((error).^2); % calculate MSE

    if MSE < bestMSE % update min error
        bestMSE = MSE;
        bestoutput=output;
        besta = a;
        bestb = b;
        bestc = c;
    end

    Et(ep) = bestMSE;

    disp(['Iteration ' num2str(ep) ': Best Cost = ' num2str(Et(ep))]);

end;

fis.output=bestoutput;


time = toc;
PlotResult(target, fis.output,'ELANFIS', time)

File: python/ELANFIS.py
import time
import numpy as np
import pandas as pd
import skfuzzy as fuzz
from PlotResult import plot_result

# Start timer
start_time = time.time()

# Load dataset
data = pd.read_csv('routput.csv').values
input_data = data[:, :-1]
target = data[:, -1]

# Initialize parameter
center, u, _, _, _, _, _ = fuzz.cluster.cmeans(
    input_data.T, 3, 2, error=1e-3, maxiter=100
)  # center = center cluster, u = membership level

samples, features = input_data.shape
n_class = 3  # total classes
ep = 0
ep_max = 1
yy, hi = np.argmax(u, axis=0), np.argmax(u, axis=0)  # hi = the class corresponding to the max value

best_mse = 1000
best_a = np.zeros((n_class, features))
best_b = np.full((n_class, features), 2)
best_c = np.zeros((n_class, features))
best_output = None

while ep < ep_max:
    ep += 1
    a = np.zeros((n_class, features))
    b = np.zeros((n_class, features))
    c = np.zeros((n_class, features))
    
    # Estimating random mf parameters
    for k in range(n_class):
        for i in range(features):
            # premise parameter: a
            Rj = np.max(input_data[:, i]) - np.min(input_data[:, i])
            m = np.sum(hi == k)
            a_temp = Rj / (2 * m - 2) if m > 1 else Rj / 2
            a_lower = a_temp * 0.5
            a_upper = a_temp * 1.5
            a[k, i] = np.random.uniform(a_lower, a_upper)
            
            # premise parameter: c
            dcc = np.random.uniform(1.9, 2.1)
            c_lower = center[k, features-1] - dcc / 2
            c_upper = center[k, features-1] + dcc / 2
            c[k, i] = np.random.uniform(c_lower, c_upper)
    
    H = []
    Mu = np.zeros((samples, n_class, features))
    
    # Calculate membership grades
    for i in range(samples):
        # Calculate firing strength
        w1 = np.ones(n_class)
        mu = np.zeros((n_class, features))
        
        for k in range(n_class):
            for j in range(features):
                mu[k, j] = 1 / (1 + ((input_data[i, j] - c[k, j]) / a[k, j]) ** (2 * b[k, j]))
                w1[k] *= mu[k, j]
                Mu[i, k, j] = mu[k, j]
        
        # Calculate Normalised Firing
        w = w1 / np.sum(w1)
        ZX = []
        
        # Generate X of f=XZ
        for k in range(n_class):
            ZX.extend(w[k] * input_data[i, :])
            ZX.append(w[k])
        
        H.append(ZX)
    
    H = np.array(H)
    
    # consequent parameter (p, q, r)
    beta = np.linalg.pinv(H) @ target  # moore pseudo inverse
    
    output = H @ beta  # calculate output from weight
    
    error = target - output
    mse = np.mean(error ** 2)  # calculate MSE
    
    if mse < best_mse:  # update min error
        best_mse = mse
        best_output = output
        best_a = a
        best_b = b
        best_c = c
    
    print(f'Iteration {ep}: Best Cost = {best_mse}')

# Create FIS output structure
fis = {'output': best_output}

elapsed_time = time.time() - start_time
plot_result(target, fis['output'], 'ELANFIS', elapsed_time)
